ggml_extra.cpp

```diff
@@ -0,0 +1,673 @@
+            auto result1 = kQuantize1(QK/2, X, L, tmpX, work, 7);
+            auto result2 = kQuantize1(QK/2, X + QK/2, L + QK/2, tmpX, work, 7);
+            auto a1fp16 = ggml_fp32_to_fp16(result1.first);
+            auto b1fp16 = ggml_fp32_to_fp16(result1.second);
+            auto a2fp16 = ggml_fp32_to_fp16(result2.first);
+            auto b2fp16 = ggml_fp32_to_fp16(result2.second);
+            std::memcpy(q, &a1fp16, sizeof(a1fp16)); q += sizeof(a1fp16);
+            std::memcpy(q, &b1fp16, sizeof(b1fp16)); q += sizeof(b1fp16);
+            std::memcpy(q, &a2fp16, sizeof(a2fp16)); q += sizeof(a2fp16);
+            std::memcpy(q, &b2fp16, sizeof(b2fp16)); q += sizeof(b2fp16);
+            for (int k=0; k<QK/2; ++k) q[k] = L[2*k] | (L[2*k+1] << 4);
+        } else {
+            auto result = type == 2 ? kQuantize1(QK, X, L, tmpX, work, 15) : kQuantize1Fast(QK, X, L, 31);
+            auto afp16 = ggml_fp32_to_fp16(result.first);
+            auto bfp16 = ggml_fp32_to_fp16(result.second);
+            std::memcpy(q, &afp16, sizeof(afp16)); q += sizeof(afp16);
+            std::memcpy(q, &bfp16, sizeof(bfp16)); q += sizeof(bfp16);
+            auto u = (uint32_t*)q;
+            *u = 0;
+            q += sizeof(uint32_t);
+            uint32_t m = 1u;
+            for (int k=0; k<QK/2; ++k) {
+                auto l1 = L[2*k], l2 = L[2*k+1];
+                if (l1 > 15) { l1 -= 16; *u |= m; }
+                m <<= 1;
+                if (l2 > 15) { l2 -= 16; *u |= m; }
+                m <<= 1;
+                q[k] = l1 | (l2 << 4);
+            }
+        }
+    };
+
+    auto bucketSize = type == 0 || type == 4 ? kBucketSize0 : kBucketSize1;
+    auto y = (char*)buffer;
+    int nchunk = (k + kChunkSize-1)/kChunkSize;
+    if (nchunk < 2) {
+        std::vector<int8_t> L(QK);
+        std::vector<std::pair<float,int>> work;
+        std::vector<float> tmpX;
+        int nb = k / QK;
+        auto x = X;
+        for (int i=0; i<nb; ++i) {
+            processOne(x, L.data(), y, work, tmpX);
+            y += bucketSize; x += QK;
+        }
+        return;
+    }
+
+    std::atomic<int> counter(0);
+    auto compute = [&counter, X, y, k, bucketSize, &processOne] () {
+        std::vector<int8_t> L(QK);
+        std::vector<std::pair<float,int>> work;
+        std::vector<float> tmpX;
+        while (true) {
+            int first = counter.fetch_add(kChunkSize, std::memory_order_relaxed);
+            if (first >= k) break;
+            int last = first + kChunkSize;
+            if (last > k) last = k;
+            auto xi = X + first;
+            auto yi = y + (first/QK)*bucketSize;
+            int n = (last - first)/QK;
+            for (int i=0; i<n; ++i) {
+                processOne(xi, L.data(), yi, work, tmpX);
+                yi += bucketSize; xi += QK;
+            }
+        }
+    };
+    int nthread = std::min(nchunk, int(std::thread::hardware_concurrency()));
+    std::vector<std::thread> workers(nthread-1);
```

Comment:
use std::jthread so you don't have to loop join them

Reply:
Currently we are targeting C++11, and std::jthread seems to be C++20.

Reply:
Actually, no. https://github.com/ggerganov/llama.cpp/blob/c12b14b77fced0ce9a0e2d81f670c3a746dec251/Makefile#L85